File: Java & Python/trees/post_order.java
/* Use 2 stacks for post order */
public class Solution {
	public ArrayList<Integer> postorderTraversal(TreeNode A) {
		Stack<TreeNode> stack1 = new Stack<>();
		Stack<TreeNode> stack2 = new Stack<>();
		ArrayList<Integer> res = new ArrayList<Integer>();
		TreeNode node;
		if (A == null)
			return null;

		stack1.push(A);
		while (!stack1.isEmpty()) {
			node = stack1.pop();
			stack2.push(node);

			if (node.left != null)
				stack1.push(node.left);
			if (node.right != null)
				stack1.push(node.right);
		}
		while (!stack2.isEmpty()) {
			node = stack2.pop();
			res.add(node.val);
		}
		return res;
	}
}

File: Java & Python/trees/post_order.py
# 2 stacks
# stack1 for node, then right, left children
# stack 2 = stack1.pop() (right child X) , then push X children onto stack 1
# [ left children -> right children ], right goes first
class Solution:
	# @param A : root node of tree
	# @return a list of integers
	def postorderTraversal(self, A):
		res, stack1, stack2 = [], [], []
		if A is None:
			return None
		stack1.append(A)
		while len(stack1) != 0:
			node = stack1.pop(0)
			stack2.insert(0, node)

			if node.left != None: 
				stack1.insert(0, node.left)
			if node.right != None:
				stack1.insert(0, node.right)
		while len(stack2) != 0:
			node = stack2.pop(0)
			res.append(node.val)
		return res


"""
2 stacks
"""